File: machine/cpp/machine/machine.cpp
#include "machine/machine.h"

#include "machine/model_registry.h"
#include <stdexcept>

namespace bmsx {
Machine::Machine(
	Memory& memoryRef,
	InputControllerInputSource& input,
	const MachineModelSpec& model
)
	: memory(memoryRef)
	, cartridgeController(memoryRef.cartridgeController())
	, irqController(memory)
	, executionAddressSpace(memory)
	, cpu(memory, irqController, executionAddressSpace)
	, scheduler(cpu)
	, audioOutput()
	, dmaController(memory, cpu, irqController, scheduler)
	, audioController(memory, audioOutput, dmaController, irqController, scheduler)
	, geometryController(memory, irqController, scheduler)
	, gxGpu(memory, cpu, irqController, scheduler, dmaController)
	, imgDecController(memory, cpu, irqController, scheduler, dmaController, model.imgDecCyclesPerOutputWord)
	, gxGte(memory, cpu, scheduler)
	, systemController(
		memory,
		cpu,
		scheduler,
		irqController,
		dmaController,
		geometryController,
		gxGpu,
		imgDecController,
		model.cpuFreqHz
	)
	, inputController(memory, input, systemController)
{
	cartridgeController.connect(memory, irqController, dmaController);
	dmaController.setTiming(
		model.dmaRamCyclesPerWord,
		model.dmaRamBurstSetupCycles,
		model.dmaSystemRomCyclesPerWord,
		model.dmaCartRomCyclesPerWord,
		model.dmaCartRomBurstSetupCycles,
		scheduler.currentNowCycles()
	);
}

void Machine::resetDevices() {
	irqController.reset();
	inputController.reset();
	dmaController.reset();
	cartridgeController.reset();
	geometryController.reset();
	gxGpu.reset();
	imgDecController.reset();
	gxGte.reset();
	audioController.reset();
	systemController.reset();
}

void Machine::refreshDeviceTimings(const MachineTiming& timing, i64 nowCycles) {
	systemController.setTiming(timing.cpuHz);
	geometryController.setTiming(timing.cpuHz, timing.geoWorkUnitsPerSec, nowCycles);
	audioController.setTiming(timing.cpuHz, nowCycles);
	gxGpu.setTiming(timing.cpuHz, nowCycles);
}

void Machine::advanceDevices(int cycles) {
	const i64 nextNow = scheduler.nowCycles() + cycles;
	geometryController.accrueCycles(cycles, nextNow);
	scheduler.advanceTo(nextNow);
}

u32 Machine::runDeviceService(uint8_t deviceKind) {
	const i64 nowCycles = scheduler.nowCycles();
	switch (deviceKind) {
		case DEVICE_SERVICE_GEO:
		geometryController.onService(nowCycles);
			return 0u;
		case DEVICE_SERVICE_DMA:
			dmaController.onService(nowCycles);
			return 0u;
		case DEVICE_SERVICE_APU:
			audioController.onService(nowCycles);
			return 0u;
		case DEVICE_SERVICE_APU_TRANSFER:
			audioController.onTransferService(nowCycles);
			return 0u;
		case DEVICE_SERVICE_GPU:
			return gxGpu.onService(nowCycles);
		case DEVICE_SERVICE_GTE:
			gxGte.onService();
			return 0u;
		case DEVICE_SERVICE_IMGDEC:
			imgDecController.onService(nowCycles);
			return 0u;
		case DEVICE_SERVICE_SYSTEM:
			systemController.onService();
			return 0u;
		default:
			throw BMSX_RUNTIME_ERROR("unknown device service kind " + std::to_string(deviceKind) + ".");
	}
}


} // namespace bmsx

File: machine/ts/machine/machine.ts
import { ApuOutputMixer } from './devices/audio/output';
import { CPU } from './cpu/cpu';
import { ExecutionAddressSpace } from './execution_address_space';
import { AudioController } from './devices/audio/controller';
import type { CartridgeController } from './devices/cartridge/controller';
import { DmaController } from './devices/dma/controller';
import { GeometryController } from './devices/geometry/controller';
import { GxGpu } from './devices/gx/gpu';
import { GxGte } from './devices/gx/gte';
import { InputController } from './devices/input/controller';
import type { InputControllerInputSource } from './devices/input/contracts';
import { ImgDecController } from './devices/imgdec/controller';
import { IrqController } from './devices/irq/controller';
import { SystemController } from './devices/system/controller';
import { Memory } from './memory/memory';
import type { MachineModelSpec } from '../spec/bmsx/model';
import {
	DEVICE_SERVICE_APU,
	DEVICE_SERVICE_APU_TRANSFER,
	DEVICE_SERVICE_DMA,
	DEVICE_SERVICE_GEO,
	DEVICE_SERVICE_GPU,
	DEVICE_SERVICE_GTE,
	DEVICE_SERVICE_IMGDEC,
	DEVICE_SERVICE_SYSTEM,
	DeviceScheduler,
} from './scheduler/device';

export type MachineTiming = {
	cpuHz: number;
	geoWorkUnitsPerSec: number;
};

export class Machine {
	public readonly cpu: CPU;
	public readonly executionAddressSpace: ExecutionAddressSpace;
	public readonly scheduler: DeviceScheduler;
	public readonly irqController: IrqController;
	public readonly systemController: SystemController;
	public readonly cartridgeController: CartridgeController;
	public readonly dmaController: DmaController;
	public readonly geometryController: GeometryController;
	public readonly gxGpu: GxGpu;
	public readonly gxGte: GxGte;
	public readonly imgDecController: ImgDecController;
	public readonly inputController: InputController;
	public readonly audioOutput: ApuOutputMixer;
	public readonly audioController: AudioController;

	public constructor(
		public readonly memory: Memory,
		input: InputControllerInputSource,
		model: MachineModelSpec,
	) {
		this.cartridgeController = this.memory.cartridgeController;
		this.irqController = new IrqController(this.memory);
		this.executionAddressSpace = new ExecutionAddressSpace(this.memory);
		this.cpu = new CPU(this.memory, this.irqController, this.executionAddressSpace);
		this.scheduler = new DeviceScheduler(this.cpu);
		this.audioOutput = new ApuOutputMixer();
		this.dmaController = new DmaController(this.memory, this.cpu, this.irqController, this.scheduler);
		this.cartridgeController.connect(this.memory, this.irqController, this.dmaController);
		this.audioController = new AudioController(this.memory, this.audioOutput, this.dmaController, this.irqController, this.scheduler);
		this.geometryController = new GeometryController(this.memory, this.irqController, this.scheduler);
		this.gxGpu = new GxGpu(
			this.memory,
			this.cpu,
			this.irqController,
			this.scheduler,
			this.dmaController,
			model.gxGpuVramBytes,
		);
		this.imgDecController = new ImgDecController(
			this.memory,
			this.cpu,
			this.irqController,
			this.scheduler,
			this.dmaController,
			model.imgDecCyclesPerOutputWord,
		);
		this.gxGte = new GxGte(this.memory, this.cpu, this.scheduler);
		this.systemController = new SystemController(
			this.memory,
			this.cpu,
			this.scheduler,
			this.irqController,
			this.dmaController,
			this.geometryController,
			this.gxGpu,
			this.imgDecController,
			model.cpuFreqHz,
		);
		this.inputController = new InputController(this.memory, input, this.systemController);
		this.dmaController.setTiming(
			model.dmaRamCyclesPerWord,
			model.dmaRamBurstSetupCycles,
			model.dmaSystemRomCyclesPerWord,
			model.dmaCartRomCyclesPerWord,
			model.dmaCartRomBurstSetupCycles,
			this.scheduler.currentNowCycles(),
		);
	}

	public resetDevices(): void {
		this.irqController.reset();
		this.inputController.reset();
		this.dmaController.reset();
		this.cartridgeController.reset();
		this.geometryController.reset();
		this.gxGpu.reset();
		this.imgDecController.reset();
		this.gxGte.reset();
		this.audioController.reset();
		this.systemController.reset();
	}

	public refreshDeviceTimings(timing: MachineTiming, nowCycles: number): void {
		this.systemController.setTiming(timing.cpuHz);
		this.geometryController.setTiming(timing.cpuHz, timing.geoWorkUnitsPerSec, nowCycles);
		this.audioController.setTiming(timing.cpuHz, nowCycles);
		this.gxGpu.setTiming(timing.cpuHz, nowCycles);
	}

	public advanceDevices(cycles: number): void {
		const nextNow = this.scheduler.nowCycles + cycles;
		this.geometryController.accrueCycles(cycles, nextNow);
		this.scheduler.advanceTo(nextNow);
	}

	public runDeviceService(deviceKind: number): number {
		const nowCycles = this.scheduler.nowCycles;
		switch (deviceKind) {
			case DEVICE_SERVICE_GEO:
				this.geometryController.onService(nowCycles);
				return 0;
			case DEVICE_SERVICE_DMA:
				this.dmaController.onService(nowCycles);
				return 0;
			case DEVICE_SERVICE_APU:
				this.audioController.onService(nowCycles);
				return 0;
			case DEVICE_SERVICE_APU_TRANSFER:
				this.audioController.onTransferService(nowCycles);
				return 0;
			case DEVICE_SERVICE_GPU:
				return this.gxGpu.onService(nowCycles);
			case DEVICE_SERVICE_GTE:
				this.gxGte.onService();
				return 0;
			case DEVICE_SERVICE_IMGDEC:
				this.imgDecController.onService(nowCycles);
				return 0;
			case DEVICE_SERVICE_SYSTEM:
				this.systemController.onService();
				return 0;
			default:
				throw new Error(`Runtime fault: unknown device service kind ${deviceKind}.`);
		}
	}

}
